File: Python3/Next Permutation.py
# 52ms, 100.0%
class Solution:
    def nextPermutation(self, nums):
        index = len(nums) - 2
        while index >= 0 and nums[index] >= nums[index + 1]:
            index -= 1
        ub = index + 1
        if index >= 0:
            for i in range(index + 2, len(nums)):
                if nums[i] > nums[index] and nums[i] <= nums[ub]:
                    ub = i
            nums[index], nums[ub] = nums[ub], nums[index]
        nums[index + 1:] = nums[index + 1:][::-1]
        

File: Java/Next Permutation.java
// 3ms, 6.48%; 40.6MB, 14.67%
class Solution {
    public static void nextPermutation(int[] nums) {
        int index = inverseLastOrderPair(nums);
        Arrays.parallelSort(nums, index + 1, nums.length);
    }

    private static int inverseLastOrderPair(int[] nums) {
        int last = nums.length - 1, cur = last;
        while (--cur >= 0) {
            for (int i = last; i > cur; --i) {
                if (nums[cur] < nums[i]) {
                    int tmp = nums[cur];
                    nums[cur] = nums[i];
                    nums[i] = tmp;
                    return cur;
                }
            }
        }
        return -1;
    }
}
